[libc][math][c23] Add f16{add,sub}f C23 math functions

overmighty opened the pull request:
Part of #93566.

overmighty committed:
[libc][math][c23] Add f16addf C23 math function

overmighty committed:
[libc][math][c23] Add f16subf C23 math function

overmighty commented:
cc @lntue

overmighty committed:
fixup! [libc][math][c23] Add f16subf C23 math function

overmighty committed:
fixup! [libc][math][c23] Add f16subf C23 math function

Fix double rounding error.

overmighty committed:
[libc][math][c23] Move f16{add,sub}f specs to llvm_libc_ext.td

overmighty commented:
Rebased to fix merge conflicts and added this commit:
- [libc][math][c23] Move f16{add,sub}f specs to llvm_libc_ext.td

overmighty commented:
Rebased and fixed the merge conflict.

lntue commented on libc/src/__support/FPUtil/generic/add_sub.h:
```diff
@@ -0,0 +1,206 @@
+  InFPBits x_bits(x);
+  InFPBits y_bits(y);
+
+  bool is_effectively_add = (x_bits.sign() == y_bits.sign()) ^ IsSub;
```
for bool, instead of `^`, use `!=`

lntue commented on libc/src/__support/FPUtil/generic/add_sub.h:
```diff
@@ -0,0 +1,206 @@
+        }
+      }
+
+      // volatile prevents Clang from converting tmp to OutType and then
```
```
if constexpr(IsSub)
  return static_cast<OutType>(-tmp);
else
  return static_cast<OutType>(tmp);
```
does not work properly without `volatile`?

overmighty replied:
I didn't think of that. Will try it.

overmighty replied:
That doesn't give the expected results either.

overmighty committed:
fixup! [libc][math][c23] Add f16subf C23 math function

Replace ^ on bools with !=.

overmighty commented:
Rebased to resolve the merge conflict and added this commit:

- **fixup! [libc][math][c23] Add f16subf C23 math function**  
  
  Replace ^ on bools with !=.
